### PR TITLE
test(v2): v2 experimental sample test infra. Part of #3505

### PR DESCRIPTION
**Pull request by Bobgy**
**Description of your changes:**
Use a KFP pipeline to run sample tests.
Part of #3505

Corresponding test config: https://github.com/GoogleCloudPlatform/oss-test-infra/pull/755

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?

    [Learn more about cherry-picking updates into the release branch](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#cherry-picking-pull-requests-to-release-branch).

### CODE REVIEW

File: v2/test/Makefile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+PROJECT=gongyuan-pipeline-test
+REPO_ROOT=../..
+GCS_ROOT=gs://$(PROJECT)/sample-test/8
+# HOST=https://48b1ee8166ba25d8-dot-asia-east1.pipelines.googleusercontent.com
+HOST=https://71a74112589c16e8-dot-asia-east1.pipelines.googleusercontent.com
+
+.PHONY: all
+all: sample-test
```
Comment (Bobgy) [inline v2/test/Makefile:8]: this is the entrypoint to run sample test

File: v2/test/sample_test.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+kaniko = kfp.components.load_component_from_file('components/kaniko.yaml')
+
+@kfp.dsl.pipeline(name='v2 sample test')
+def v2_sample_test(
+    context: 'URI' = CONTEXT,
+    launcher_destination: 'URI' = 'gcr.io/gongyuan-pipeline-test/kfp-launcher',
+    gcs_root: 'URI' = 'gs://gongyuan-test/v2',
+    samples_destination: 'URI' = 'gcr.io/gongyuan-pipeline-test/v2-sample-test',
+    kfp_host: 'URI' = 'http://ml-pipeline:8888',
+    samples_config: 'JSONArray' = json.dumps([
```
Comment (Ark-kun) [inline v2/test/sample_test.py:23]: JFYI: For KFP it's OK to just write it like this:
`samples_config: list = [...],`
The `list` type is converted to `JsonArray` and the `json.dumps` serialization is used.

File: v2/test/components/kaniko.yaml [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+- {name: context_sub_path, type: Path, default: ""}
+- {name: destination, type: URI, description: destination should not contain tag}
+outputs:
+- {name: digest, type: URI, description: Image URI with full digest}
+metadata:
+  annotations:
+    author: Yuan Gong <gongyuan94@gmail.com>
+implementation:
+  container:
+    # The debug image contains busybox, while the normal images do not.
```
Comment (Ark-kun) [inline v2/test/components/kaniko.yaml:25]: Great find. I've hit this problem as well when creating a Kaniko component for the tutorials.

File: v2/test/sample_test.py (hunk 2 above)
Comment (Bobgy) [reply v2/test/sample_test.py:23]: nice, I'll do it instead

**Comment by Ark-kun** [comment]
Thank you for test infra improvement, Yuan!
/lgtm

**Comment by Ark-kun** [comment]
Do we need to check in the `context.tar.gz` file? What's inside?

**Comment by Bobgy** [comment]
@Ark-kun it's a package of the source code folder, see Makefile

> 	mkdir -p tmp
	tar -C $(REPO_ROOT) \
		--exclude-vcs-ignores \
		-zcvf tmp/context.tar.gz \
		sdk v2

**Comment by Bobgy** [comment]
TODOs:
* [ ] use built launcher image in tests
* [ ] separate pipeline verification into its own file
* [ ] prepare presubmit script
* [ ] learn from v1 sample test infra to see if anything in addition can help
* [ ] allow migrating v1 samples to run on this test infra

**Pull request merged**